contracts/dezswap_pair/src/contract.rs

```diff
@@ -309,11 +314,15 @@ pub fn provide_liquidity(
         };
 
         let remain_amount = deposits[i] - desired_amount;
+        refund_assets.push(Asset {
+            info: pool.info.clone(),
+            amount: remain_amount,
+        });
 
         if let AssetInfo::NativeToken { denom, .. } = &pool.info {
             if !remain_amount.is_zero() {
                 messages.push(CosmosMsg::Bank(BankMsg::Send {
-                    to_address: info.sender.to_string(),
+                    to_address: refund_receiver.to_string(),
                     amount: coins(remain_amount.u128(), denom),
                 }))
             }
```

Comment:
shouldn't `recipient` be updated as well? (line:334)

Reply:
Line 334 is msg to get cw20 from sender.

It can be confusing, but when providing liquidity, native_token sends the token to the contract from the sender, and the contract brings the cw20 from the sender.

Reply:
yup but isn't it a bit weird that native tokens are refunded to `refund_receiver` but cw token is not?
not sure which way would be better. anyway since we record refunded assets below whatever the token's type is so I thought it would go to `refund_receiver`.

Reply:
cw20 is brought when msg is successful, so there is no need to refund. I agree that the if statement and the else if statement are confusing because they do completely different things.